hw/xbox/mcpx_apu.c

```diff
@@ -644,28 +796,122 @@ static void se_frame(void *opaque)
+      mixbin[0][i] = (rand() % 0x10000 - 0x8000) << 8;
+    }
+
+    /* Write VP results to the GP DSP MIXBUF */
+    for(unsigned int i = 0; i < NUM_MIXBINS; i++) {
+        for(unsigned int j = 0; j < NUM_SAMPLES_PER_FRAME; j++) {
+            dsp_write_memory(d->gp.dsp, 'X', 3072 + i * 0x20 + j,
+                             mixbin[i][j] & 0xFFFFFF);
+        }
+#if 1
+        if (i < 6) {
+            gp_scratch_rw(d, mixbin[i],
+                          0x8000 /*+ d->xbox_ds_gp_audio_offset*/ + i * 0x800,
+                          32 * 4, 1);
+        }
+#endif
+    }
+
+    /* Kickoff DSP processing */
     if ((d->gp.regs[NV_PAPU_GPRST] & NV_PAPU_GPRST_GPRST)
         && (d->gp.regs[NV_PAPU_GPRST] & NV_PAPU_GPRST_GPDSPRST)) {
         dsp_start_frame(d->gp.dsp);
 
         // hax
-        dsp_run(d->gp.dsp, 1000);
+        dsp_run(d->gp.dsp, 50);
     }
     if ((d->ep.regs[NV_PAPU_EPRST] & NV_PAPU_GPRST_GPRST)
         && (d->ep.regs[NV_PAPU_EPRST] & NV_PAPU_GPRST_GPDSPRST)) {
         dsp_start_frame(d->ep.dsp);
 
         // hax
-        // dsp_run(d->ep.dsp, 1000);
+        //dsp_run(d->ep.dsp, 1000);
     }
+
+    //FIXME: The following should not be necessary. It's a hack.
+    //
+    //       Normally, the GP mixes the MIXBINs, then DMAs the mixed result to
+    //       the EP, which then encodes AC3 and DMAs to the ACI.
+    //       So the final mixing of MIXBINs is typically done on GP.
+    //
+    //       As we don't have the working DSP / DMA / ACI, we need our own
+    //       mixing and output for now.
+    //       In the future, this code can be removed.
+
+#if ROUTE_VP_AUDIO
+
+    /* Route all VP MIXBIN contents equally to downmixed mono output */
+    route_vp_audio(d, mixbin);
+
+#endif
+
+#if ROUTE_XBOX_DS_GP_AUDIO
+
+    /* Route stereo audio that GP wants to DMA to EP to our own output */
+    route_xbox_ds_gp_audio(d);
+
+#endif
+
 }
 
+/* TODO: this should be on a thread so it waits on the voice lock */
+static void process_all_se_frames(void *opaque)
+{
+    MCPXAPUState *d = opaque;
+    timer_mod(d->se.frame_timer, qemu_clock_get_ms(QEMU_CLOCK_VIRTUAL) + 10);
+    MCPX_DPRINTF("mcpx frame ping\n");
+
+    static uint64_t last_ns = 0;
+    uint64_t current_ns = qemu_clock_get_ns(QEMU_CLOCK_REALTIME);
+
+    // We output multiple frames, because timing of this routine is bad.
+    // It's 10000.0us, when it should be 666.66...us
+    uint64_t elapsed_time = current_ns - last_ns;
+    uint64_t frame_time = 666666;
+    unsigned int frames = elapsed_time / frame_time;
+
+    unsigned int max_frames = 20;
+    if (frames > max_frames) {
+      frames = max_frames;
+    }
+
+    printf("Time passed: %llu ns\n", elapsed_time);
+    printf("Doing %d frames; %d samples; %d bytes\n", frames, frames * NUM_SAMPLES_PER_FRAME, frames * NUM_SAMPLES_PER_FRAME * 4);
+    last_ns = current_ns;
+
+frames = 1;
+    for(unsigned int frame = 0; frame < frames; frame++) {
+        se_frame(opaque);
+    }
+}
```

Comment:
This is a hack... I'm not sure how it ended up in this branch? This was part of the DSP testing